File: php/exception_handling.php
<?php
// Exception Handling in PHP

// 1. Basic Exception Handling
function divide($a, $b) {
    try {
        if ($b == 0) {
            throw new Exception("Division by zero");
        }
        return $a / $b;
    } catch (Exception $e) {
        echo "Caught exception: " . $e->getMessage() . "\n";
    } finally {
        echo "Execution completed.\n";
    }
}

// Example usage
echo divide(10, 2); // Outputs 5
echo divide(10, 0); // Outputs error message

// 2. Custom Exception Handling
class CustomException extends Exception {
    public function errorMessage() {
        return "Custom error on line " . $this->getLine() . " in " . $this->getFile() . ": " . $this->getMessage();
    }
}

function testException($value) {
    try {
        if ($value < 0) {
            throw new CustomException("Negative value not allowed");
        }
        echo "Value is $value\n";
    } catch (CustomException $e) {
        echo $e->errorMessage() . "\n";
    }
}

// Example usage
testException(10); // Outputs "Value is 10"
testException(-5); // Outputs custom error message

// 3. Multiple Exceptions
function multiCatchExample($a, $b) {
    try {
        if ($b == 0) {
            throw new DivisionByZeroError("Cannot divide by zero");
        }
        if (!is_numeric($a) || !is_numeric($b)) {
            throw new InvalidArgumentException("Both arguments must be numbers");
        }
        return $a / $b;
    } catch (DivisionByZeroError $e) {
        echo "Error: " . $e->getMessage() . "\n";
    } catch (InvalidArgumentException $e) {
        echo "Error: " . $e->getMessage() . "\n";
    } finally {
        echo "Finished execution.\n";
    }
}

// Example usage
echo multiCatchExample(10, 2); // Outputs 5
echo multiCatchExample(10, 0); // Outputs division by zero error
echo multiCatchExample(10, 'a'); // Outputs invalid argument error

// 4. Using Finally for Cleanup
function openFile($filename) {
    try {
        $file = fopen($filename, "r");
        if (!$file) {
            throw new Exception("File not found");
        }
        echo fread($file, filesize($filename));
    } catch (Exception $e) {
        echo "Caught exception: " . $e->getMessage() . "\n";
    } finally {
        if (isset($file) && $file) {
            fclose($file);
            echo "File closed.\n";
        }
    }
}

// Example usage
openFile("example.txt"); // Outputs file content or error
?>

File: python/exception_handling.py
# exception_handling.py

# Exception Handling in Python

# 1. Basic Exception Handling
from contextlib import contextmanager


def divide(a, b):
    try:
        result = a / b
    except ZeroDivisionError:
        print("Error: Cannot divide by zero.")
    except TypeError:
        print("Error: Invalid input type.")
    else:
        print(f"Result: {result}")
    finally:
        print("Execution completed.")


# Example usage of basic exception handling
divide(10, 2)
divide(10, 0)
divide(10, "a")


# 2. Custom Exceptions
class CustomError(Exception):
    """Base class for other exceptions"""

    pass


class NegativeValueError(CustomError):
    """Raised when the input value is negative"""

    pass


def check_positive(value):
    if value < 0:
        raise NegativeValueError("Negative value is not allowed.")
    else:
        print(f"Value is positive: {value}")


# Example usage of custom exception
try:
    check_positive(10)
    check_positive(-5)
except NegativeValueError as e:
    print(e)


# 3. Handling Multiple Exceptions in One Block
def multiple_exceptions(a, b):
    try:
        result = a / b
    except (ZeroDivisionError, TypeError) as e:
        print(f"Error: {e}")
    else:
        print(f"Result: {result}")
    finally:
        print("Execution completed.")


# Example usage of multiple exception handling
multiple_exceptions(10, 2)
multiple_exceptions(10, 0)
multiple_exceptions(10, "b")


# 4. Raising Exceptions
def raise_exception_example(value):
    if not isinstance(value, int):
        raise TypeError("Only integers are allowed.")
    if value < 0:
        raise ValueError("Negative integers are not allowed.")
    print(f"Value is: {value}")


# Example usage of raising exceptions
try:
    raise_exception_example(10)
    raise_exception_example(-5)
    raise_exception_example("10")
except (TypeError, ValueError) as e:
    print(f"Caught an exception: {e}")


# 5. Using `finally` for Cleanup Actions
def file_handling_example(filename):
    try:
        file = open(filename, "r")
        data = file.read()
        print(data)
    except FileNotFoundError:
        print(f"Error: {filename} not found.")
    finally:
        if "file" in locals():
            file.close()
            print(f"File {filename} closed.")


# Example usage of finally for cleanup
file_handling_example("example.txt")
file_handling_example("non_existent.txt")

# 6. Context Managers and Exception Handling


@contextmanager
def managed_resource(resource):
    try:
        print(f"Acquiring resource: {resource}")
        yield resource
    except Exception as e:
        print(f"Error while using resource: {e}")
    finally:
        print(f"Releasing resource: {resource}")


# Example usage of context manager with exception handling
with managed_resource("ResourceName") as res:
    print(f"Using {res}")
    raise ValueError("An error occurred while using the resource.")
